sphinx_gallery/gen_rst.py

```diff
@@ -135,36 +138,25 @@ def write(self, data):
     `Generated by Sphinx-Gallery <https://sphinx-gallery.readthedocs.io>`_\n"""
 
 
-def codestr2rst(codestr, lang='python'):
+def codestr2rst(codestr, lang='python', lineno=None):
     """Return reStructuredText code block from code string"""
-    code_directive = "\n.. code-block:: {0}\n\n".format(lang)
+    if lineno is not None:
+        if LooseVersion(sphinx.__version__) >= '1.3':
+            # Sphinx only starts numbering from the first non-empty line.
+            blank_lines = codestr.count('\n', 0, -len(codestr.lstrip()))
+            lineno = '   :lineno-start: {0}\n'.format(lineno + blank_lines)
+        else:
+            lineno = '   :linenos:\n'
+    else:
+        lineno = ''
+    code_directive = "\n.. code-block:: {0}\n{1}\n".format(lang, lineno)
     indented_block = indent(codestr, ' ' * 4)
     return code_directive + indented_block
 
 
-def extract_thumbnail_number(text):
-    """ Pull out the thumbnail image number specified in the docstring. """
-
-    # check whether the user has specified a specific thumbnail image
-    pattr = re.compile(
-        r"^\s*#\s*sphinx_gallery_thumbnail_number\s*=\s*([0-9]+)\s*$",
-        flags=re.MULTILINE)
-    match = pattr.search(text)
-
-    if match is None:
-        # by default, use the first figure created
-        thumbnail_number = 1
-    else:
-        thumbnail_number = int(match.groups()[0])
-
-    return thumbnail_number
-
-
-def extract_intro(filename):
+def extract_intro(filename, docstring):
```

Comment:
I feel this is a good opportunity to change the logic of how Sphinx-Gallery works. This function fails after the example has been built if the docstring is not compliant, and crashes the entire gallery build process.
Then one has to change the file to update the docstring and rebuild the file.
I would like to put the check of a compliant docstring in the `get_docstring_and_rest` function and raise the exception there.